src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -267,33 +265,62 @@ minRequiredLength switch
+                switch (code.FindOptimizations.FindMode)
                 {
-                    writer.WriteLine("return true;");
+                    case FindNextStartingPositionMode.LeadingPrefix_LeftToRight_CaseSensitive:
+                        Debug.Assert(!string.IsNullOrEmpty(code.FindOptimizations.LeadingCaseSensitivePrefix));
+                        EmitIndexOf_LeftToRight(code.FindOptimizations.LeadingCaseSensitivePrefix);
+                        break;
+
+                    case FindNextStartingPositionMode.LeadingPrefix_RightToLeft_CaseSensitive:
+                        Debug.Assert(!string.IsNullOrEmpty(code.FindOptimizations.LeadingCaseSensitivePrefix));
+                        EmitIndexOf_RightToLeft(code.FindOptimizations.LeadingCaseSensitivePrefix);
+                        break;
+
+                    case FindNextStartingPositionMode.FixedSets_LeftToRight_CaseSensitive:
+                    case FindNextStartingPositionMode.FixedSets_LeftToRight_CaseInsensitive:
+                    case FindNextStartingPositionMode.LeadingSet_LeftToRight_CaseSensitive:
+                    case FindNextStartingPositionMode.LeadingSet_LeftToRight_CaseInsensitive:
+                        Debug.Assert(code.FindOptimizations.FixedDistanceSets is { Count: > 0 });
+                        EmitFixedSet_LeftToRight();
+                        break;
+
+                    case FindNextStartingPositionMode.LeadingSet_RightToLeft_CaseSensitive:
+                    case FindNextStartingPositionMode.LeadingSet_RightToLeft_CaseInsensitive:
+                        Debug.Assert(code.FindOptimizations.FixedDistanceSets is { Count: > 0 });
+                        EmitFixedSet_RightToLeft();
+                        break;
+
+                    // Already emitted earlier
+                    case FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Beginning:
+                    case FindNextStartingPositionMode.LeadingAnchor_LeftToRight_End:
+                    case FindNextStartingPositionMode.LeadingAnchor_LeftToRight_EndZ:
+                    case FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Start:
+                    case FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Beginning:
+                    case FindNextStartingPositionMode.LeadingAnchor_RightToLeft_End:
+                    case FindNextStartingPositionMode.LeadingAnchor_RightToLeft_EndZ:
+                    case FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Start:
+                        Debug.Assert(code.FindOptimizations.LeadingAnchor != 0);
+                        goto case FindNextStartingPositionMode.NoSearch;
+
+                    default:
+                        Debug.Fail($"Unexpected mode: {code.FindOptimizations.FindMode}");
+                        goto case FindNextStartingPositionMode.NoSearch;
+
+                    case FindNextStartingPositionMode.NoSearch:
```

Comment:
I think I found an issue testing this locally. Basically the test pattern I'm using is:
`@"^https://api\.github\.com/repos/dotnet/runtime/issues/([0-9]+)$"`

What seems to happen now, is that for that pattern, when emitting FindFirstChar, we are first adding a `return true;` after making the check of `if (runtextpos > runtextbeg)`, but after that we are no longer emitting the IndexOf code we were doing, and instead now FindMode is set to `LeadingAnchor_LeftToRight_Beginning` which is calling a `goto case .. NoSearch` which eventually is adding a second `return true;` to the method which is incorrect. Is this a problem when selecting the FileMode here? Given this is just a long Multi, shouldn't it have selected IndexOf as the FindMode?

Reply:
Actually I understand now a bit more, it does seem that we don't want the IndexOf here. so the first return true is being emitted because my expression has a leading anchor, so all that should be checked is to ensure that we are at the beginning of the input. I think the only thing we need to fix here, is that line 319 shouldn't be going to NoSearch and instead just breaking, as all the cases when you have a leading anchor, we would have already called EmitAnchors() method which would already have added a `return true;`

Reply:
Exactly.  This code used to exist outside of a helper, and the return statements in it would short-circuit the remaining code.  When I moved it into a helper, the return statements stopped having the same effect, such that we'd proceed to emit (dead) code after these returns.